transformer_engine/pytorch/attention.py

```diff
@@ -515,22 +516,41 @@ class AttnFuncWithCP(torch.autograd.Function):
 
     @staticmethod
     def forward(ctx, is_training, q, k, v, cu_seqlens_q, cu_seqlens_k, max_seqlen_q, max_seqlen_k,
-                dropout_p, cp_group, cp_global_ranks, cp_stream, softmax_scale, attn_mask_type,
-                deterministic, use_fused_attention):
+                dropout_p, cp_group, cp_global_ranks, cp_stream, softmax_scale, qkv_format,
+                attn_mask_type, attn_bias_type, attn_bias, deterministic, use_fused_attention):
         if softmax_scale is None:
             softmax_scale = q.shape[-1] ** (-0.5)
 
         cp_size = get_distributed_world_size(cp_group)
         rank = get_distributed_rank(cp_group)
         send_dst = cp_global_ranks[(rank + 1) % cp_size]
-        recv_src = cp_global_ranks[(rank + cp_size - 1) % cp_size]
+        recv_src = cp_global_ranks[(rank - 1) % cp_size]
         batch_p2p_comm = int(os.getenv("NVTE_BATCH_MHA_P2P_COMM", "0")) or (cp_size == 2)
 
         causal = (attn_mask_type == "causal")
 
+        qkv_layout = qkv_format + "_" + qkv_format + "_" + qkv_format
```

Comment:
We use `qkv_layout` in `FusedAttention.forward()` and `FlashAttention.forward()`, and here we are converting it again back to `qkv_format`. Would it be cleaner if we used `qkv_layout` here?

Also, does CP currently only support contiguous `q/k/v` cases, i.e. `qkv_layout=bshd_bshd_bshd or sbhd_sbhd_sbhd`? What if `DotProductAttention` passes down `q/k/v` tensors in `bshd_bs2hd` layout?

Reply:
Yes, CP only can work with `bshd_bshd_bshd` and `sbhd_sbhd_sbhd`. This is because CP has communication of KV. With received KV, qkv_layout can change. For example, if a GPU has qkv_layout of `bs3hd` for its local qkv. There is no communication for Q, while it does attention compute with local Q and received KV from another GPU, qkv_layout is not `bs3hd` any more. Hence, qkv_layout hacking is not quite simple, let me think about this. I will submit another PR after I figure it out.

However, your review really reminded me that there is a bug in my code. I made [this commit](https://github.com/NVIDIA/TransformerEngine/pull/726/commits/8e2776c1df25633096d14ebe3bcae7e44e617582) to make sure QKV always have layout of `bshd_bshd_bshd` or `sbhd_sbhd_sbhd`. FlashAttention did it, but I missed it in FusedAttention. Thanks.